R/refineCenters.R

```diff
@@ -0,0 +1,143 @@
+#' @title Refine Peak Centers
+#'
+#' @description
+#' This function refines the center values of a peak by weighting the y values
+#' in the neightbourhood that belong most likely to the same peak.
+#'
+#' @param x `numeric`, i.e. m/z values.
+#' @param y `numeric`, i.e. intensity values.
+#' @param p `numeric`, peak centers.
+#' @param k `integer(1)`, number of values left and right of the peak that
+#'  should be considered in the weighted mean calculation.
+#' @param threshold `double`, ratio of the maximal peak intensity. Just values
+#'  above are used for the weighted mean calclulation.
+#' @param descending `logical`, if `TRUE` just values between the nearest
+#'  valleys around the peak centers are used.
+#'
+#' @details
+#' For `descending = FALSE` the function looks for the `k` nearest neighbouring
+#' data points and use their `x` for weighted mean with their corresponding `y`
+#' values as weights for calculation of the new peak center. If `k` are chosen
+#' too large it could result in skewed peak centers, see example below.
+#' If `descending = TRUE` is used the `k` should be general larger because it is
+#' trimmed automatically to the nearest valleys on both sides of the peak so the
+#' problem with skewed centers is rare.
+#'
+#' @author Sebastian Gibb, Johannes Rainer
+#' @export
+#' @examples
+#' ints <- c(5, 8, 12, 7, 4, 9, 15, 16, 11, 8, 3, 2, 3, 9, 12, 14, 13, 8, 3)
+#' mzs <- seq_along(ints)
+#'
+#' plot(mzs, ints, type = "h")
+#'
+#' pidx <- c(3, 8, 16)
+#' points(mzs[pidx], ints[pidx], pch = 16)
+#'
+#' ## Use the weighted average considering the adjacent mz
+#' mzs1 <- refineCenters(mzs, ints, pidx,
+#'                       k = 2L, descending = FALSE, threshold = 0)
+#' mzs2 <- refineCenters(mzs, ints, pidx,
+#'                       k = 5L, descending = FALSE, threshold = 0)
+#' mzs3 <- refineCenters(mzs, ints, pidx,
+#'                       k = 5L, descending = TRUE, threshold = 0)
+#' points(mzs1, ints[pidx], col = "red", type = "h")
+#' ## please recognize the artificial moved center of the first peak caused by a
+#' ## too large k, here
+#' points(mzs2, ints[pidx], col = "blue", type = "h")
+#' points(mzs3, ints[pidx], col = "green", type = "h")
+#' legend("topright",
+#'        legend = paste0("k = ", c(2, 5, 5),
+#'                        ", descending =", c("FALSE", "FALSE", "TRUE")),
+#'        col = c("red", "blue", "green"), lwd = 1)
+refineCenters <- function(x, y, p, k = 2, threshold = 0.33, descending = TRUE) {
+    if (!is.numeric(x))
+        stop("'x' has to be a numeric vector.")
+    if (!is.numeric(y) || length(x) != length(y))
```

Comment:
Why not having also `!is.numeric(x)` here and remove the extra `if (!is.numeric(x))` above?

Reply:
Don't know. Adopted to your suggestion.